File: conversions/query-string/elide-query-string.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-query-string (URL Query String Parser)
#
# Demonstrates calling the TypeScript query-string implementation from Ruby
# for consistent URL parameter handling across microservices.

puts "=== Ruby Calling TypeScript Query String Parser ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# qs = Elide.require('./elide-query-string.ts')

# Example 1: Parse URL query string
# query = "?name=Alice&role=admin&active=true&score=95.5"
# params = qs.parse(query)
# puts "Query: #{query}"
# puts "Parsed: #{params.inspect}"
# # Output: {:name => "Alice", :role => "admin", :active => true, :score => 95.5}

# Example 2: Stringify parameters
# params = {search: 'ruby elide', page: 2, limit: 50}
# query_string = qs.stringify(params)
# puts "Parameters: #{params.inspect}"
# puts "Query string: #{query_string}"
# # Output: search=ruby%20elide&page=2&limit=50

# Example 3: Handle arrays (consistent with Node.js!)
# query = "tags=ruby&tags=rails&tags=sinatra"
# parsed = qs.parse(query)
# puts "Array query: #{query}"
# puts "Parsed: #{parsed.inspect}"
# # Output: {:tags => ["ruby", "rails", "sinatra"]}

# Example 4: Sinatra API with consistent query parsing
# require 'sinatra'
# require 'json'
#
# get '/api/products' do
#   content_type :json
#
#   # Parse query string same way as Node.js Express!
#   params = qs.parse(request.query_string)
#
#   category = params['category']
#   brands = params['brands']  # Array support
#   min_price = params['minPrice']  # Number parsed
#   max_price = params['maxPrice']
#
#   products = Product.where(category: category)
#                     .where(brand: brands)
#                     .where('price >= ? AND price <= ?', min_price, max_price)
#
#   products.to_json
# end

# Example 5: Rails controller
# class ProductsController < ApplicationController
#   def index
#     # Instead of Rails params:
#     # qs_params = QS.parse(request.query_string)
#     #
#     # Same parsing as Node.js and Python APIs!
#     # @products = Product.filter_by(qs_params)
#     #
#     # render json: @products
#   end
# end

# Example 6: HTTP client with consistent params
# require 'net/http'
# require 'uri'
#
# def call_api_service(endpoint, params)
#   # Build query string using same logic as Node.js client
#   query = qs.stringify(params, {arrayFormat: 'bracket', sort: true})
#   url = URI("https://api.example.com/#{endpoint}?#{query}")
#
#   response = Net::HTTP.get_response(url)
#   JSON.parse(response.body)
# end
#
# # Usage
# results = call_api_service('search', {
#   q: 'elide',
#   filters: ['polyglot', 'runtime'],
#   page: 1,
#   limit: 20
# })

# Example 7: Sidekiq background job with API calls
# class DataSyncWorker
#   include Sidekiq::Worker
#
#   def perform(sync_params)
#     # Build query for external API
#     query = qs.stringify(sync_params, {arrayFormat: 'comma'})
#     url = "https://partner-api.com/data?#{query}"
#
#     # Same query format as Node.js worker!
#     response = HTTParty.get(url)
#     process_data(response.parsed_response)
#   end
# end

puts "Real-world use case:"
puts "- Ruby Sinatra API reads query: ?brands[]=nike&brands[]=adidas&page=1"
puts "- Uses same query-string parser as Node.js Express API"
puts "- Guarantees identical parameter parsing across entire platform"
puts ""

puts "Example: Microservices Parameter Consistency"
puts "┌──────────────────────────────────────┐"
puts "│  Elide Query String (TypeScript)    │"
puts "│  conversions/query-string/           │"
puts "└──────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌─────────┐ ┌─────────┐ ┌─────────┐"
puts "    │ Node.js │ │  Ruby   │ │ Python  │"
puts "    │ Express │ │ Sinatra │ │ Flask   │"
puts "    └─────────┘ └─────────┘ └─────────┘"
puts "     GET /api/search?q=shoes&brands[]=nike&brands[]=adidas"
puts "     All three parse identically:"
puts "     { q: 'shoes', brands: ['nike', 'adidas'] }"
puts "     ✓ Zero inconsistency!"
puts ""

puts "Problem Solved:"
puts "Before: Each language parses query strings differently"
puts "  - Node.js: qs.parse() → custom behavior"
puts "  - Ruby: Rack::Utils.parse_nested_query() → different array handling"
puts "  - Python: urllib.parse.parse_qs() → always returns lists"
puts "  - Result: Same query string = different parsed results = BUGS"
puts ""
puts "After: One Elide implementation = identical parsing everywhere"
puts "  - All languages use same TypeScript parser"
puts "  - Same options, same behavior, same results"
puts "  - Perfect API parameter consistency across services"
puts ""

puts "Configuration Example:"
puts "  # config/query_parsing.yml (shared by all services)"
puts "  arrayFormat: 'bracket'   # brands[]=a&brands[]=b"
puts "  parseNumbers: true       # page=1 → number"
puts "  parseBooleans: true      # active=true → boolean"
puts "  sort: true               # alphabetical keys"
puts "  "
puts "  # All microservices use identical settings:"
puts "  params = qs.parse(query_string, config)"
puts ""

puts "Use Cases:"
puts "  ✓ REST API parameter parsing (Sinatra, Rails, Rack)"
puts "  ✓ Microservice URL building (consistent with Node.js/Python)"
puts "  ✓ Search/filter query handling (arrays, nested objects)"
puts "  ✓ HTTP client query strings (API calls to other services)"
puts "  ✓ Background job parameters (Sidekiq with API integration)"
puts ""

File: conversions/query-string/elide-query-string.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-query-string (URL Query String Parser)

Demonstrates calling the TypeScript query-string implementation from Python
for consistent URL parameter handling across all services.

Benefits:
- One query string parser shared across TypeScript and Python
- Consistent API parameter parsing across services
- No need for urllib.parse.parse_qs quirks
- Unified array and nested object handling
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# from elide import require
# qs = require('./elide-query-string.ts')

print("=== Python Calling TypeScript Query String Parser ===\n")

# Example 1: Parse URL query string
# query = "?name=John&age=30&city=NYC&active=true"
# params = qs.parse(query)
# print(f"Query: {query}")
# print(f"Parsed: {params}")
# # Output: {'name': 'John', 'age': 30, 'city': 'NYC', 'active': True}

# Example 2: Stringify parameters
# params = {'search': 'elide runtime', 'page': 1, 'limit': 20}
# query_string = qs.stringify(params)
# print(f"Parameters: {params}")
# print(f"Query string: {query_string}")
# # Output: search=elide%20runtime&page=1&limit=20

# Example 3: Handle arrays (consistent with Node.js!)
# query = "tags=python&tags=typescript&tags=polyglot"
# parsed = qs.parse(query)
# print(f"Array query: {query}")
# print(f"Parsed: {parsed}")
# # Output: {'tags': ['python', 'typescript', 'polyglot']}

# Example 4: Flask API with consistent query parsing
# from flask import Flask, request
# app = Flask(__name__)
#
# @app.route('/api/search')
# def search():
#     # Instead of Flask's request.args:
#     # params = qs.parse(request.query_string.decode())
#     #
#     # Same parsing as Node.js Express API!
#     # search_term = params.get('q')
#     # filters = params.get('filters')  # Array support
#     # page = params.get('page', 1)     # Number parsed
#     #
#     # return jsonify(results=perform_search(search_term, filters, page))

# Example 5: Django views
# from django.http import JsonResponse
#
# def product_list(request):
#     # Parse query parameters identically to Node.js service
#     # params = qs.parse(request.META['QUERY_STRING'])
#     #
#     # category = params.get('category')
#     # brands = params.get('brands')  # Arrays work!
#     # min_price = params.get('minPrice')
#     # max_price = params.get('maxPrice')
#     #
#     # products = Product.objects.filter(
#     #     category=category,
#     #     brand__in=brands,
#     #     price__gte=min_price,
#     #     price__lte=max_price
#     # )
#     #
#     # return JsonResponse({'products': list(products.values())})

# Example 6: API client with consistent params
# import requests
#
# def call_microservice(endpoint, params):
#     # Build query string using same logic as Node.js client
#     query = qs.stringify(params, {'arrayFormat': 'bracket', 'sort': True})
#     url = f"https://api.example.com/{endpoint}?{query}"
#     response = requests.get(url)
#     return response.json()
#
# # Usage
# results = call_microservice('search', {
#     'q': 'elide',
#     'filters': ['polyglot', 'runtime'],
#     'page': 1,
#     'limit': 20
# })

print("Real-world use case:")
print("- Python Flask API reads query: ?filters=python&filters=java&page=1")
print("- Uses same query-string parser as Node.js Express API")
print("- Guarantees identical parameter parsing across entire platform")
print()

print("Example: API Parameter Consistency")
print("┌──────────────────────────────────────┐")
print("│  Elide Query String (TypeScript)    │")
print("│  conversions/query-string/           │")
print("└──────────────────────────────────────┘")
print("         ↓                    ↓")
print("    ┌─────────┐          ┌─────────┐")
print("    │ Node.js │          │ Python  │")
print("    │ Express │          │ Flask   │")
print("    └─────────┘          └─────────┘")
print("     GET /api/search?q=elide&tags[]=js&tags[]=py")
print("     Both parse identically:")
print("     { q: 'elide', tags: ['js', 'py'] }")
print("     ✓ Perfect consistency!")
print()

print("Problem Solved:")
print("Before: Node.js + Python parse query strings differently")
print("  - Node.js: qs.parse('tags=a&tags=b') → {tags: ['a', 'b']}")
print("  - Python: parse_qs('tags=a&tags=b') → {'tags': ['a', 'b']}  # List always!")
print("  - Different defaults, different behavior = BUGS")
print()
print("After: One Elide implementation = identical parsing")
print("  - Both languages use same TypeScript parser")
print("  - Same options, same behavior, same results")
print("  - Zero query parameter inconsistency bugs")
print()

print("Configuration Example:")
print("  # API Gateway config (shared by all services)")
print("  query_parsing:")
print("    arrayFormat: 'bracket'  # tags[]=a&tags[]=b")
print("    parseNumbers: true      # page=1 → number")
print("    parseBooleans: true     # active=true → boolean")
print("  ")
print("  # Both Node.js and Python use identical settings:")
print("  params = qs.parse(queryString, config)")
print()

print("Use Cases:")
print("  ✓ REST API parameter parsing (Flask, FastAPI, Django)")
print("  ✓ Microservice URL building (consistent with Node.js)")
print("  ✓ Search query handling (arrays, filters, pagination)")
print("  ✓ Form data processing (web scraping, automation)")
print("  ✓ API testing (verify query parameter handling)")
print()
